### ISSUE TITLE
M269 TMA notebooks not processed

### ISSUE THREAD

**Issue by mwermelinger** [dsa-ou/allowed#24]
With the initial `%%javascript` code cell, `allowed` doesn't report disallowed constructs, only cells with syntax errors. Without that cell, `allowed` works correctly.

The issue is thus due to the cell magic or to the fact that the metadata is `trusted: false`.

**Comment by densnow** [comment]
It seems that when using the `transform_cell` function on the concatenated source, the function would interpret the whole lot as a cell hence with a cell magic it was all added as an argument to the `run_cell_magic()` .  

The solution is to transform the source of each cell before adding to the `source_list` at least that is working for me, I will do further testing and submit the PR.

**Comment by mwermelinger** [comment]
Indeed the problem is that we concatenate all cells into one, and so `transform_cell` produces different results depending on where the cell magic appears. If we start the notebook with a cell
```
%%javascript
var head=document.getElementsByTagName('head')[0]
```
then the result is 
```python
get_ipython().run_cell_magic('javascript', '', 
'var head=document.getElementsByTagName(\'head\')[0]\n...')
```
where `...` is the content of the following code cells. The code isn't checked because it's in a string passed to a function.

If the same cell comes after a syntactically correct Python cell, then the resulting transformation is 
```python
... # previous cell's Python code
get_ipython().run_line_magic('%javascript', '')
var head=document.getElementsByTagName('head')[0]
... # next cell's Python code
```
Since the second line is Javascript, not Python, a syntax error ensues and the whole code isn't checked. 

The current `sample.ipynb` works fine because `%%capture output` appears in the last cell, so it's transformed as `get_ipython().run_line_magic('%capture', 'output')` and the rest of the code can be checked as normal.

If we transform each cell before concatenating them, then a cell magic 'gobbles' the whole cell and it won't be checked.

Maybe the best solution is indeed to replace IPython commands by 'no-ops' like `print(42)` (although magics like `%timeit` in assignments are trickier to handle) and let non-Python code lead to syntax error messages.

For the moment, let's simply transform each cell. That's enough to handle the only use of cell_magics in M269: `%%javascript` in the assessment.

**Comment by densnow** [comment]
I was thinking about replacing cell magics only and let `transform_cell` handle the line magics.

For the cell magic to work in Jupyter notebooks, it needs to be on the first line in the cell on its own with valid arguments (as far as I can tell: the usage documentation is sparse). This would make it relatively straightforward to check and replace. It would be nice to check cells with cell magics for the cases when they have valid Python, but of course it is not urgent as you say.

There are another couple of  small changes that could be made to the code that I might as well mention here before I forget:

in line [549](https://github.com/dsa-ou/allowed/blob/75635a05d58fabd147930c66e7e8fdd96810afbe/allowed.py#L549) of `allowed.py` we have
```python
 for cell_line_num in range(1, len(cell_source.split("\n")) + 1):
```
This could be changed to
```python
 for cell_line_num in range(1, cell_source.count("\n")) + 2):
```
so a list does not have to be created each time.

The other thing  I added that is not quite right is in `sample.ipynb`: the markdown above  code cell 3 (the cell magic) says that the rest of the code will be added to the second parameter, It seems to be the third parameter, so that needs amending.

**Comment by mwermelinger** [comment]
Fixed by #25
